File: cross-contract-hello-as/contract/assembly/external.ts
import { context, ContractPromise, ContractPromiseResult, u128 } from "near-sdk-as";

// Constants
export const TGAS: u64 = 1000000000000;
export const NO_DEPOSIT: u128 = u128.Zero;
export const XCC_SUCCESS = 1

// Auxiliary Method: Make the callback private and return its result
export function get_callback_result(): ContractPromiseResult {
  assert(
    context.predecessor == context.contractName,
    "Only the contract itself can call this method"
  );

  // Return the result from the external pool
  const results = ContractPromise.getResults();
  assert(results.length == 1, "This is a callback method");
  return results[0];
}

// Auxiliary classes for cross-contract calls
@nearBindgen
export class GreetingArgs {
  constructor(
    public message: string
  ) { }
}

@nearBindgen
export class CbArgs {
  constructor(
    public donor: string,
    public amount: u128
  ) { }
}

@nearBindgen
export class ValidatorArgs {
  constructor(public account_id: string) {}
}

File: cross-contract-hello-rs/contract/src/external.rs
use near_sdk::{ext_contract};

pub const TGAS: u64 = 1_000_000_000_000;
pub const NO_DEPOSIT: u128 = 0;
pub const XCC_SUCCESS: u64 = 1;

// Interface of this contract, for callbacks
#[ext_contract(this_contract)]
trait Callbacks {
  fn query_greeting_callback(&mut self) -> String;
  fn change_greeting_callback(&mut self) -> bool;
}

// Validator interface, for cross-contract calls
#[ext_contract(hello_near)]
trait HelloNear {
  #[payable]
  fn get_greeting(&mut self) -> String;
  fn set_greeting(&self, message: String);
}